libraries/bot-azure/src/main/java/com/microsoft/bot/azure/blobs/BlobsTranscriptStore.java

```diff
@@ -0,0 +1,423 @@
+                        }
+                    }
+                    throw ex;
+                }
+            }
+        }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    private CompletableFuture<Activity> getActivityFromBlobClient(BlobClient blobClient) {
+        ByteArrayOutputStream content = new ByteArrayOutputStream();
+        blobClient.download(content);
+        String contentString = new String(content.toByteArray());
+        try {
+            return jacksonAdapter.deserialize(contentString, Activity.class);
+        } catch (IOException exception) {
+            return CompletableFuture.completedFuture(null);
+        }
+    }
+
+    private CompletableFuture<Void> logActivityToBlobClient(Activity activity, BlobClient blobClient,
+                                                            Boolean overwrite) {
+        if (overwrite == null) {
+            overwrite = false;
+        }
+        String activityJson = null;
+        try {
+            activityJson = jacksonAdapter.serialize(activity);
+        } catch (IOException exception) {
+            exception.printStackTrace();
+        }
+        InputStream data = new ByteArrayInputStream(activityJson.getBytes(StandardCharsets.UTF_8));
+
+        //verify the corresponding length
+        blobClient.upload(data, 4l, overwrite);
+
+        Map<String, String> metaData = new HashMap<String, String>();
+        metaData.put("Id", activity.getId());
+        if (activity.getFrom() != null) {
+            metaData.put("FromId", activity.getFrom().getId());
+        }
+
+        if (activity.getRecipient() != null) {
+            metaData.put("RecipientId", activity.getRecipient().getId());
+        }
+        metaData.put("Timestamp", activity.getTimestamp().toString());
+
+        blobClient.setMetadata(metaData);
+
+        return CompletableFuture.completedFuture(null);
+    }
+
+    private String getBlobName(Activity activity) {
+        String blobName = String.format("%1$s/%2$s/%3$s-%4$s.json",
+            sanitizeKey(activity.getChannelId()), activity.getConversation().getId(),
+            this.formatTicks(activity.getTimestamp()), sanitizeKey(activity.getId()));
+
+        return blobName;
+    }
+
+    private String sanitizeKey(String key) {
+        // Blob Name rules: case-sensitive any url char
+        try {
+            return URLEncoder.encode(key, StandardCharsets.UTF_8.name());
+        } catch (Exception ex) {
+            return key;
+        }
+    }
+
+    /**
+     * Formats a timestamp in a way that is consistent with the C# SDK.
+     * @param dateTime
+     * @return
+     */
+    private String formatTicks(OffsetDateTime dateTime) {
+        final long epochTicks = 621355968000000000L; // the number of .net ticks at the unix epoch
+        final int ticksPerMillisecond = 10000;
+        final long ticks = epochTicks + dateTime.toInstant().toEpochMilli() * ticksPerMillisecond;
+        return String.valueOf(ticks);
+    }
```

Comment:
Are we sure about this? What's the `x` in [C#](https://github.com/microsoft/botbuilder-dotnet/blob/main/libraries/Microsoft.Bot.Builder.Azure.Blobs/BlobsTranscriptStore.cs#L402)?